### PR TITLE
[Bug 1158909] Add l10n extraction and the scaffolding for localization in JS and Nunjucks templates.

### PR DESCRIPTION
**Pull request by mythmon**
This PR does several things:

* Adds functions to JS (`gettext`, and `ngettext`) and Nunjucks templates (`_` and `_plural`) to call into the localization system.
* Adds stubs for those functions until real localization is ready.
* Changes the Nunjucks rendering strategy to use an environment. This way globals can be added to the rendering context (such as `_` and `_plural`).
* Building an extraction task for use with Grunt that is compatible with the way Kitsune wants template files.
* Add gettext calls to all strings that should be localized in JS and Nunjucks files.
* Some clean up.

Things this does not do:

* Provide a solution for localizing all the strings current in plain html files. I suggest the parts of these files that contain localizable strings be converted into Nunjucks templates.
* Pull the localized strings back from Kitsune.
* Actually localize anything (only stubbed versions of `gettext` and friends are available). This will depend a lot on how we pull these strings back from Kitsune.
* Remove the bits of the l20n system that is in place for date formatting. 

r?

### CODE REVIEW

**Comment by mythmon** [comment]
Any updates here?

File: app/js/l10n.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
+(function() {
+  function format(string, formatKwargs) {
+    if (formatKwargs === undefined) {
+      return string;
+    }
+
+    return string.replace(/{[^}]+}/, function(input) {
```
Comment (Rik) [inline app/js/l10n.js]: This should be `/{[^}]+}/g` to match several arguments. I've noticed this when you click on a top helper, the numbers of questions solved and helpful votes displays as {n}.

File: package.json [hunk 2]
```diff
@@ -10,14 +10,19 @@
   "license": "MIT",
   "dependencies": {},
   "devDependencies": {
+    "acorn": "^1.1.0",
+    "es6-promises": "^1.0.10",
```
Comment (Rik) [inline package.json]: We already added `es6-promise` (not the s lacking) in https://github.com/mozilla/buddyup/commit/029f6014cd3eadb0043e175d8769522008c7d1f8#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R13. Could we update this pull request to use that nom module instead?

File: tasks/extract.js [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+                      '\n' + err.toString());
+    }
+
+    return parseTree.findAll(nunjucksNodes.FunCall)
+    .filter(function(node) {
+      // Exclude functions calls that aren't to gettext.
+      return (
+        node.name &&
+        node.name instanceof nunjucksNodes.Symbol &&
+        ['_', '_plural', 'gettext', 'ngettext'].indexOf(node.name.value) !== -1
```
Comment (Rik) [inline tasks/extract.js]: Shouldn't this be `['_', '_plural']` only? From `l10n.js`, we're not adding `gettext` and `ngettext` to the `nunjucksEnv`.

**Comment by Rik** [comment]
I've fixed the nits and pushed this to the https://github.com/mozilla/buddyup/tree/l10n branch

**Comment by mythmon** [comment]
When I did the extract+merge steps into Kitsune's locale repo, I learned there are a few syntax errors in the generated files. I'll file a new pull request to fix this in the next few days.

**Pull request closed without merging**